test/e2e/apps/deployment.go

```diff
@@ -416,7 +413,9 @@ var _ = SIGDescribe("Deployment", func() {
 			case watch.Modified:
 				if deployment, ok := event.Object.(*appsv1.Deployment); ok {
 					found := deployment.ObjectMeta.Name == testDeployment.Name &&
-						deployment.ObjectMeta.Labels["test-deployment-static"] == "true"
+						deployment.Status.ReadyReplicas == testDeploymentNoReplicas &&
+						deployment.Status.AvailableReplicas == testDeploymentAvailableReplicas
+
```

Comment:
If we don't get `deploymentsList` just before `patching the DeploymentStatus`, `deploymentsList` will still be the one here

https://github.com/kubernetes/kubernetes/blob/247555c2b793f4e2be8a0882f7bc12a0a1da83b2/test/e2e/apps/deployment.go#L319

then the event we get will be the event k8s did after `updating the Deployment`.

After `updating the Deployment`, k8s will modify `deployment.Status.ReadyReplicas` and `deployment.Status.AvailableReplicas` to 0 first.

Reply:
Starting the watch operation with an earlier `ResourceVersion` is a problem.

https://github.com/kubernetes/kubernetes/blob/247555c2b793f4e2be8a0882f7bc12a0a1da83b2/test/e2e/apps/deployment.go#L411-L425

In line `419`, the code return with `found`, which is the event deployment controller did after `updating the Deployment` before `patching the DeploymentStatus`. But what we expected is the event `patching the DeploymentStatus`.